diff --git a/src/class/hid/hid_host.c b/src/class/hid/hid_host.c
index eef584d74..a3cc7d6d7 100644
--- a/src/class/hid/hid_host.c
+++ b/src/class/hid/hid_host.c
@@ -410,7 +410,7 @@ bool tuh_hid_send_report(uint8_t daddr, uint8_t idx, uint8_t report_id, const vo
     ++len; // 1 more byte for report_id
   }
 
-  TU_LOG3_MEM(p_hid->epout_buf, len, 2);
+  TU_LOG3_MEM(epbuf->epout, len, 2);
 
   if (!usbh_edpt_xfer(daddr, p_hid->ep_out, epbuf->epout, len)) {
     usbh_edpt_release(daddr, p_hid->ep_out);
@@ -445,7 +445,7 @@ bool hidh_xfer_cb(uint8_t daddr, uint8_t ep_addr, xfer_result_t result, uint32_t
 
   if (dir == TUSB_DIR_IN) {
     TU_LOG_DRV("  Get Report callback (%u, %u)\r\n", daddr, idx);
-    TU_LOG3_MEM(p_hid->epin_buf, xferred_bytes, 2);
+    TU_LOG3_MEM(epbuf->epin, xferred_bytes, 2);
     tuh_hid_report_received_cb(daddr, idx, epbuf->epin, (uint16_t) xferred_bytes);
   } else {
     if (tuh_hid_report_sent_cb) {
diff --git a/src/class/vendor/vendor_host.c b/src/class/vendor/vendor_host.c
deleted file mode 100644
index e66c5007f..000000000
--- a/src/class/vendor/vendor_host.c
+++ /dev/null
@@ -1,146 +0,0 @@
-/*
- * The MIT License (MIT)
- *
- * Copyright (c) 2019 Ha Thach (tinyusb.org)
- *
- * Permission is hereby granted, free of charge, to any person obtaining a copy
- * of this software and associated documentation files (the "Software"), to deal
- * in the Software without restriction, including without limitation the rights
- * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
- * copies of the Software, and to permit persons to whom the Software is
- * furnished to do so, subject to the following conditions:
- *
- * The above copyright notice and this permission notice shall be included in
- * all copies or substantial portions of the Software.
- *
- * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
- * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
- * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
- * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
- * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
- * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
- * THE SOFTWARE.
- *
- * This file is part of the TinyUSB stack.
- */
-
-#include "tusb_option.h"
-
-#if (CFG_TUH_ENABLED && CFG_TUH_VENDOR)
-
-//--------------------------------------------------------------------+
-// INCLUDE
-//--------------------------------------------------------------------+
-#include "host/usbh.h"
-#include "vendor_host.h"
-
-//--------------------------------------------------------------------+
-// MACRO CONSTANT TYPEDEF
-//--------------------------------------------------------------------+
-
-//--------------------------------------------------------------------+
-// INTERNAL OBJECT & FUNCTION DECLARATION
-//--------------------------------------------------------------------+
-custom_interface_info_t custom_interface[CFG_TUH_DEVICE_MAX];
-
-static tusb_error_t cush_validate_paras(uint8_t dev_addr, uint16_t vendor_id, uint16_t product_id, void * p_buffer, uint16_t length)
-{
-  if ( !tusbh_custom_is_mounted(dev_addr, vendor_id, product_id) )
-  {
-    return TUSB_ERROR_DEVICE_NOT_READY;
-  }
-
-  TU_ASSERT( p_buffer != NULL && length != 0, TUSB_ERROR_INVALID_PARA);
-
-  return TUSB_ERROR_NONE;
-}
-//--------------------------------------------------------------------+
-// APPLICATION API (need to check parameters)
-//--------------------------------------------------------------------+
-tusb_error_t tusbh_custom_read(uint8_t dev_addr, uint16_t vendor_id, uint16_t product_id, void * p_buffer, uint16_t length)
-{
-  TU_ASSERT_ERR( cush_validate_paras(dev_addr, vendor_id, product_id, p_buffer, length) );
-
-  if ( !hcd_pipe_is_idle(custom_interface[dev_addr-1].pipe_in) )
-  {
-    return TUSB_ERROR_INTERFACE_IS_BUSY;
-  }
-
-  (void) usbh_edpt_xfer( custom_interface[dev_addr-1].pipe_in, p_buffer, length);
-
-  return TUSB_ERROR_NONE;
-}
-
-tusb_error_t tusbh_custom_write(uint8_t dev_addr, uint16_t vendor_id, uint16_t product_id, void const * p_data, uint16_t length)
-{
-  TU_ASSERT_ERR( cush_validate_paras(dev_addr, vendor_id, product_id, p_data, length) );
-
-  if ( !hcd_pipe_is_idle(custom_interface[dev_addr-1].pipe_out) )
-  {
-    return TUSB_ERROR_INTERFACE_IS_BUSY;
-  }
-
-  (void) usbh_edpt_xfer( custom_interface[dev_addr-1].pipe_out, p_data, length);
-
-  return TUSB_ERROR_NONE;
-}
-
-//--------------------------------------------------------------------+
-// USBH-CLASS API
-//--------------------------------------------------------------------+
-void cush_init(void)
-{
-  tu_memclr(&custom_interface, sizeof(custom_interface_info_t) * CFG_TUH_DEVICE_MAX);
-}
-
-tusb_error_t cush_open_subtask(uint8_t dev_addr, tusb_desc_interface_t const *p_interface_desc, uint16_t *p_length)
-{
-  // FIXME quick hack to test lpc1k custom class with 2 bulk endpoints
-  uint8_t const *p_desc = (uint8_t const *) p_interface_desc;
-  p_desc = tu_desc_next(p_desc);
-
-  //------------- Bulk Endpoints Descriptor -------------//
-  for(uint32_t i=0; i<2; i++)
-  {
-    tusb_desc_endpoint_t const *p_endpoint = (tusb_desc_endpoint_t const *) p_desc;
-    TU_ASSERT(TUSB_DESC_ENDPOINT == p_endpoint->bDescriptorType, TUSB_ERROR_INVALID_PARA);
-
-    pipe_handle_t * p_pipe_hdl =  ( p_endpoint->bEndpointAddress &  TUSB_DIR_IN_MASK ) ?
-                         &custom_interface[dev_addr-1].pipe_in : &custom_interface[dev_addr-1].pipe_out;
-    *p_pipe_hdl = usbh_edpt_open(dev_addr, p_endpoint, TUSB_CLASS_VENDOR_SPECIFIC);
-    TU_ASSERT ( pipehandle_is_valid(*p_pipe_hdl), TUSB_ERROR_HCD_OPEN_PIPE_FAILED );
-
-    p_desc = tu_desc_next(p_desc);
-  }
-
-  (*p_length) = sizeof(tusb_desc_interface_t) + 2*sizeof(tusb_desc_endpoint_t);
-  return TUSB_ERROR_NONE;
-}
-
-void cush_isr(pipe_handle_t pipe_hdl, xfer_result_t event)
-{
-
-}
-
-void cush_close(uint8_t dev_addr)
-{
-  tusb_error_t err1, err2;
-  custom_interface_info_t * p_interface = &custom_interface[dev_addr-1];
-
-  // TODO re-consider to check pipe valid before calling pipe_close
-  if( pipehandle_is_valid( p_interface->pipe_in ) )
-  {
-    err1 = hcd_pipe_close( p_interface->pipe_in );
-  }
-
-  if ( pipehandle_is_valid( p_interface->pipe_out ) )
-  {
-    err2 = hcd_pipe_close( p_interface->pipe_out );
-  }
-
-  tu_memclr(p_interface, sizeof(custom_interface_info_t));
-
-  TU_ASSERT(err1 == TUSB_ERROR_NONE && err2 == TUSB_ERROR_NONE, (void) 0 );
-}
-
-#endif
diff --git a/src/class/vendor/vendor_host.h b/src/class/vendor/vendor_host.h
deleted file mode 100644
index acfebe7a4..000000000
--- a/src/class/vendor/vendor_host.h
+++ /dev/null
@@ -1,67 +0,0 @@
-/*
- * The MIT License (MIT)
- *
- * Copyright (c) 2019 Ha Thach (tinyusb.org)
- *
- * Permission is hereby granted, free of charge, to any person obtaining a copy
- * of this software and associated documentation files (the "Software"), to deal
- * in the Software without restriction, including without limitation the rights
- * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
- * copies of the Software, and to permit persons to whom the Software is
- * furnished to do so, subject to the following conditions:
- *
- * The above copyright notice and this permission notice shall be included in
- * all copies or substantial portions of the Software.
- *
- * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
- * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
- * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
- * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
- * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
- * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
- * THE SOFTWARE.
- *
- * This file is part of the TinyUSB stack.
- */
-
-#ifndef _TUSB_VENDOR_HOST_H_
-#define _TUSB_VENDOR_HOST_H_
-
-#include "common/tusb_common.h"
-
-#ifdef __cplusplus
- extern "C" {
-#endif
-
-typedef struct {
-  pipe_handle_t pipe_in;
-  pipe_handle_t pipe_out;
-}custom_interface_info_t;
-
-//--------------------------------------------------------------------+
-// USBH-CLASS DRIVER API
-//--------------------------------------------------------------------+
-static inline bool tusbh_custom_is_mounted(uint8_t dev_addr, uint16_t vendor_id, uint16_t product_id)
-{
-  (void) vendor_id; // TODO check this later
-  (void) product_id;
-//  return (tusbh_device_get_mounted_class_flag(dev_addr) & TU_BIT(TUSB_CLASS_MAPPED_INDEX_END-1) ) != 0;
-  return false;
-}
-
-bool tusbh_custom_read(uint8_t dev_addr, uint16_t vendor_id, uint16_t product_id, void * p_buffer, uint16_t length);
-bool tusbh_custom_write(uint8_t dev_addr, uint16_t vendor_id, uint16_t product_id, void const * p_data, uint16_t length);
-
-//--------------------------------------------------------------------+
-// Internal Class Driver API
-//--------------------------------------------------------------------+
-void cush_init(void);
-bool cush_open_subtask(uint8_t dev_addr, tusb_desc_interface_t const *p_interface_desc, uint16_t *p_length);
-void cush_isr(pipe_handle_t pipe_hdl, xfer_result_t event);
-void cush_close(uint8_t dev_addr);
-
-#ifdef __cplusplus
- }
-#endif
-
-#endif /* _TUSB_VENDOR_HOST_H_ */
diff --git a/src/host/usbh.c b/src/host/usbh.c
index a2994cde7..8e045a39c 100644
--- a/src/host/usbh.c
+++ b/src/host/usbh.c
@@ -199,18 +199,6 @@ static usbh_class_driver_t const usbh_class_drivers[] = {
       .close      = hub_close
   },
   #endif
-
-  #if CFG_TUH_VENDOR
-  {
-    .name       = DRIVER_NAME("VENDOR"),
-    .init       = cush_init,
-    .deinit     = cush_deinit,
-    .open       = cush_open,
-    .set_config = cush_set_config,
-    .xfer_cb    = cush_isr,
-    .close      = cush_close
-  }
-  #endif
 };
 
 enum { BUILTIN_DRIVER_COUNT = TU_ARRAY_SIZE(usbh_class_drivers) };
diff --git a/src/portable/synopsys/dwc2/dcd_dwc2.c b/src/portable/synopsys/dwc2/dcd_dwc2.c
index c461d9a79..afc1de421 100644
--- a/src/portable/synopsys/dwc2/dcd_dwc2.c
+++ b/src/portable/synopsys/dwc2/dcd_dwc2.c
@@ -437,16 +437,24 @@ bool dcd_init(uint8_t rhport, const tusb_rhport_init_t* rh_init) {
   gahbcfg |= GAHBCFG_GINT; // Enable global interrupt
   dwc2->gahbcfg = gahbcfg;
 
+  dwc2_int_alloc(rhport, TUSB_ROLE_DEVICE);
+
   dcd_connect(rhport);
   return true;
 }
 
+bool dcd_deinit(uint8_t rhport)
+{
+  dwc2_int_free(rhport);
+  return true;
+}
+
 void dcd_int_enable(uint8_t rhport) {
-  dwc2_dcd_int_enable(rhport);
+  dwc2_int_set(rhport, TUSB_ROLE_DEVICE, true);
 }
 
 void dcd_int_disable(uint8_t rhport) {
-  dwc2_dcd_int_disable(rhport);
+  dwc2_int_set(rhport, TUSB_ROLE_DEVICE, false);
 }
 
 void dcd_set_address(uint8_t rhport, uint8_t dev_addr) {
diff --git a/src/portable/synopsys/dwc2/dwc2_esp32.h b/src/portable/synopsys/dwc2/dwc2_esp32.h
index 3309760ff..e110f6350 100644
--- a/src/portable/synopsys/dwc2/dwc2_esp32.h
+++ b/src/portable/synopsys/dwc2/dwc2_esp32.h
@@ -32,6 +32,9 @@
  extern "C" {
 #endif
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+
 #include "freertos/FreeRTOS.h"
 #include "freertos/task.h"
 
@@ -39,6 +42,11 @@
 #include "soc/periph_defs.h"
 #include "soc/usb_wrap_struct.h"
 
+#pragma GCC diagnostic pop
+
+void dcd_int_handler(uint8_t);
+void hcd_int_handler(uint8_t, bool);
+
 #if TU_CHECK_MCU(OPT_MCU_ESP32S2, OPT_MCU_ESP32S3)
 #define DWC2_FS_REG_BASE   0x60080000UL
 #define DWC2_EP_MAX        7
@@ -80,17 +88,23 @@ static void dwc2_int_handler_wrap(void* arg) {
 #endif
 }
 
-TU_ATTR_ALWAYS_INLINE static inline void dwc2_int_set(uint8_t rhport, tusb_role_t role, bool enabled) {
-  if (enabled) {
+static inline void dwc2_int_alloc(uint8_t rhport, tusb_role_t role) {
     esp_intr_alloc(_dwc2_controller[rhport].irqnum, ESP_INTR_FLAG_LOWMED,
-                   dwc2_int_handler_wrap, (void*)(uintptr_t)tu_u16(role, rhport), &usb_ih[rhport]);
-  } else {
+                    dwc2_int_handler_wrap, (void*)(uintptr_t)tu_u16(role, rhport), &usb_ih[rhport]);
+}
+
+static inline void dwc2_int_free(uint8_t rhport) {
     esp_intr_free(usb_ih[rhport]);
-  }
+    usb_ih[rhport] = NULL;
 }
 
-#define dwc2_dcd_int_enable(_rhport)  dwc2_int_set(_rhport, TUSB_ROLE_DEVICE, true)
-#define dwc2_dcd_int_disable(_rhport) dwc2_int_set(_rhport, TUSB_ROLE_DEVICE, false)
+TU_ATTR_ALWAYS_INLINE static inline void dwc2_int_set(uint8_t rhport, tusb_role_t, bool enabled) {
+  if (enabled) {
+      esp_intr_enable(usb_ih[rhport]);
+  } else {
+      esp_intr_disable(usb_ih[rhport]);
+  }
+}
 
 TU_ATTR_ALWAYS_INLINE static inline void dwc2_remote_wakeup_delay(void) {
   vTaskDelay(pdMS_TO_TICKS(1));
diff --git a/src/portable/synopsys/dwc2/hcd_dwc2.c b/src/portable/synopsys/dwc2/hcd_dwc2.c
index ebbb6200a..19a3367e8 100644
--- a/src/portable/synopsys/dwc2/hcd_dwc2.c
+++ b/src/portable/synopsys/dwc2/hcd_dwc2.c
@@ -388,9 +388,19 @@ bool hcd_init(uint8_t rhport, const tusb_rhport_init_t* rh_init) {
   gahbcfg |= GAHBCFG_GINT;   // Enable global interrupt
   dwc2->gahbcfg = gahbcfg;
 
+  dwc2_int_alloc(rhport, TUSB_ROLE_HOST);
+
+  return true;
+}
+
+
+bool hcd_deinit(uint8_t rhport)
+{
+  dwc2_int_free(rhport);
   return true;
 }
 
+
 // Enable USB interrupt
 void hcd_int_enable (uint8_t rhport) {
   dwc2_int_set(rhport, TUSB_ROLE_HOST, true);
@@ -1260,8 +1270,11 @@ static void handle_hprt_irq(uint8_t rhport, bool in_isr) {
 
     if (hprt_bm.conn_status) {
       hcd_event_device_attach(rhport, in_isr);
+      dwc2->gintsts = GINTSTS_DISCINT;
+      dwc2->gintmsk |= GINTSTS_DISCINT;
     } else {
       hcd_event_device_remove(rhport, in_isr);
+      dwc2->gintmsk &= ~GINTSTS_DISCINT;
     }
   }
 
@@ -1312,6 +1325,11 @@ void hcd_int_handler(uint8_t rhport, bool in_isr) {
     }
   }
 
+  if (gintsts & GINTSTS_DISCINT) {
+      hcd_event_device_remove(rhport, in_isr);
+      dwc2->gintmsk &= ~GINTSTS_DISCINT;
+  }
+
   if (gintsts & GINTSTS_HPRTINT) {
     // Host port interrupt: source is cleared in HPRT register
     // TU_LOG1_HEX(dwc2->hprt);
diff --git a/src/portable/template/dcd_template.c b/src/portable/template/dcd_template.c
index 3738ac0cb..5d85b20b7 100644
--- a/src/portable/template/dcd_template.c
+++ b/src/portable/template/dcd_template.c
@@ -112,6 +112,15 @@ void dcd_edpt_close_all (uint8_t rhport) {
   (void) rhport;
 }
 
+void dcd_edpt_close(uint8_t rhport, uint8_t ep_addr) {
+  (void) rhport;
+  (void) ep_addr;
+}
+
+void dcd_int_handler(uint8_t rhport) {
+  (void) rhport;
+}
+
 // Submit a transfer, When complete dcd_event_xfer_complete() is invoked to notify the stack
 bool dcd_edpt_xfer (uint8_t rhport, uint8_t ep_addr, uint8_t * buffer, uint16_t total_bytes) {
   (void) rhport;
diff --git a/src/tusb.c b/src/tusb.c
index 85ab1d6ae..df4d9589c 100644
--- a/src/tusb.c
+++ b/src/tusb.c
@@ -41,20 +41,6 @@
 
 tusb_role_t _tusb_rhport_role[TUP_USBIP_CONTROLLER_NUM] = { TUSB_ROLE_INVALID };
 
-//--------------------------------------------------------------------
-// Weak/Default API, can be overwritten by Application
-//--------------------------------------------------------------------
-
-TU_ATTR_WEAK void tusb_time_delay_ms_api(uint32_t ms) {
-#if CFG_TUSB_OS != OPT_OS_NONE
-  osal_task_delay(ms);
-#else
-  // delay using millis() (if implemented) and/or frame number if possible
-  const uint32_t time_ms = tusb_time_millis_api();
-  while ((tusb_time_millis_api() - time_ms) < ms) {}
-#endif
-}
-
 //--------------------------------------------------------------------+
 // Public API
 //--------------------------------------------------------------------+
diff --git a/src/tusb.h b/src/tusb.h
index cb6021b33..8cf62be81 100644
--- a/src/tusb.h
+++ b/src/tusb.h
@@ -58,10 +58,6 @@
   #if CFG_TUH_CDC
     #include "class/cdc/cdc_host.h"
   #endif
-
-  #if CFG_TUH_VENDOR
-    #include "class/vendor/vendor_host.h"
-  #endif
 #else
   #ifndef tuh_int_handler
   #define tuh_int_handler(...)
